x-pack/plugins/fleet/server/services/metrics/fetch_agent_metrics.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { CoreSetup } from '@kbn/core/server';
+
+import { AGENTS_INDEX } from '../../../common';
+
+import type { AgentPerVersion, AgentUsage } from '../../collectors/agent_collectors';
+import { getAgentUsage } from '../../collectors/agent_collectors';
+import { getInternalClients } from '../../collectors/helpers';
+import { appContextService } from '../app_context';
+
+export interface AgentMetrics {
+  agents: AgentUsage;
+  agents_per_version: AgentPerVersion[];
+  upgrading_step: UpgradingSteps;
+}
+
+export interface UpgradingSteps {
+  requested: number;
+  scheduled: number;
+  downloading: number;
+  extracting: number;
+  replacing: number;
+  restarting: number;
+  watching: number;
+  rollback: number;
+  failed: number;
+}
+
+export const fetchAgentMetrics = async (
+  core: CoreSetup,
+  abortController: AbortController
+): Promise<AgentMetrics | undefined> => {
+  const [soClient, esClient] = await getInternalClients(core);
+  if (!soClient || !esClient) {
+    return;
+  }
+  const usage = {
+    agents: await getAgentUsage(soClient, esClient),
+    agents_per_version: await getAgentsPerVersion(esClient, abortController),
+    upgrading_step: await getUpgradingSteps(esClient, abortController),
+  };
+  return usage;
+};
+
+export const getAgentsPerVersion = async (
+  esClient: ElasticsearchClient,
+  abortController: AbortController
+): Promise<AgentPerVersion[]> => {
+  try {
+    const response = await esClient.search(
```

Comment:
Is it worth wrapping these in our `retryTransientEsErrors` helper?

Reply:
added